Mathlib/Control/Fold.lean

```diff
@@ -0,0 +1,475 @@
+```
+with
+```
+f : α → β → α
+x : α
+[y₀,y₁] : List β
+```
+
+We can view the above as a composition of functions:
+```
+... = f (f x y₀) y₁              : rfl
+... = flip f y₁ (flip f y₀ x)    : rfl
+... = (flip f y₁ ∘ flip f y₀) x  : rfl
+```
+
+We can use traverse and const to construct this composition:
+```
+calc   const.run (traverse (λ y, const.mk' (flip f y)) [y₀,y₁]) x
+     = const.run ((::) <$> const.mk' (flip f y₀) <*> traverse (λ y, const.mk' (flip f y)) [y₁]) x
+...  = const.run ((::) <$> const.mk' (flip f y₀) <*>
+         ( (::) <$> const.mk' (flip f y₁) <*> traverse (λ y, const.mk' (flip f y)) [] )) x
+...  = const.run ((::) <$> const.mk' (flip f y₀) <*>
+         ( (::) <$> const.mk' (flip f y₁) <*> pure [] )) x
+...  = const.run ( ((::) <$> const.mk' (flip f y₁) <*> pure []) ∘
+         ((::) <$> const.mk' (flip f y₀)) ) x
+...  = const.run ( const.mk' (flip f y₁) ∘ const.mk' (flip f y₀) ) x
+...  = const.run ( flip f y₁ ∘ flip f y₀ ) x
+...  = f (f x y₀) y₁
+```
+
+And this is how `const` turns a monoid into an applicative functor and
+how the monoid of endofunctions define `Foldl`.
+-/
+@[reducible]
+def Foldl (α : Type u) : Type u :=
+  (End α)ᵐᵒᵖ
+#align monoid.foldl Monoid.Foldl
+
+-- porting note: no docstring present in mathlib3
+@[nolint docBlame]
+def Foldl.mk (f : α → α) : Foldl α :=
+  op f
+#align monoid.foldl.mk Monoid.Foldl.mk
+
+-- porting note: no docstring present in mathlib3
+@[nolint docBlame]
+def Foldl.get (x : Foldl α) : α → α :=
+  unop x
+#align monoid.foldl.get Monoid.Foldl.get
+
+-- porting note: no docstring present in mathlib3
+@[simps, nolint docBlame]
+def Foldl.ofFreeMonoid (f : β → α → β) : FreeMonoid α →* Monoid.Foldl β
+    where
+  toFun xs := op <| flip (List.foldl f) (FreeMonoid.toList xs)
+  map_one' := rfl
+  map_mul' := by
+    intros; simp only [FreeMonoid.toList_mul, flip, unop_op, List.foldl_append, op_inj]; rfl
+#align monoid.foldl.of_free_monoid Monoid.Foldl.ofFreeMonoid
+
+-- porting note: no docstring present in mathlib3
+@[reducible, nolint docBlame]
+def Foldr (α : Type u) : Type u :=
+  End α
+#align monoid.foldr Monoid.Foldr
+
+-- porting note: no docstring present in mathlib3
+@[nolint docBlame]
+def Foldr.mk (f : α → α) : Foldr α :=
+  f
+#align monoid.foldr.mk Monoid.Foldr.mk
+
+-- porting note: no docstring present in mathlib3
+@[nolint docBlame]
+def Foldr.get (x : Foldr α) : α → α :=
+  x
+#align monoid.foldr.get Monoid.Foldr.get
+
+-- porting note: no docstring present in mathlib3
+@[simps, nolint docBlame]
+def Foldr.ofFreeMonoid (f : α → β → β) : FreeMonoid α →* Monoid.Foldr β
+    where
+  toFun xs := flip (List.foldr f) (FreeMonoid.toList xs)
+  map_one' := rfl
+  map_mul' _ _ := funext fun _ => List.foldr_append _ _ _ _
+#align monoid.foldr.of_free_monoid Monoid.Foldr.ofFreeMonoid
+
+-- porting note: no docstring present in mathlib3
+@[reducible, nolint docBlame]
+def Mfoldl (m : Type u → Type u) [Monad m] (α : Type u) : Type u :=
```

Comment:
Monadic functions follow a different naming scheme now: the `M` goes at the end, so this should be `foldlM`. Can you please rename this and the others below?